working-with-openai/open-ai-api-in-projects.md

```diff
@@ -0,0 +1,381 @@
+
+    characters = db.session.scalars(character_query)
+    response = []
+
+    for character in characters:
+        response.append(
+            {
+                "id" : character.id,
+                "name" : character.name,
+                "personality" : character.personality,
+                "occupation" : character.occupation,
+                "age" : character.age
+            }
+        )
+
+    return jsonify(response)
+
+@bp.get("/<char_id>/greetings")
+def get_greetings(char_id):
+    pass
+
+@bp.post("/<char_id>/generate")
+def add_greetings(char_id):
+    pass
+
+def validate_model(cls,id):
+    try:
+        id = int(id)
+    except:
+        response =  response = {"message": f"{cls.__name__} {id} invalid"}
+        abort(make_response(response , 400))
+
+    query = db.select(cls).where(cls.id == id)
+    model = db.session.scalar(query)
+    if model:
+        return model
+
+    response = {"message": f"{cls.__name__} {id} not found"}
+    abort(make_response(response, 404))
+```
+</details>
+</br>
+
+## Set up and connect the Database
+
+Create and connect your database using the steps below
+
+1. Set up a development database in psql called npc_generator_dev
+2. Create a .env file and add the database as your `SQLALCHEMY_DATABASE_URI`
+3. Run `flask init`, `flask migrate` and `flask upgrade` to connect the API to your database.
+4. Test the POST and GET methods for `Character` in Postman.
+
+<br />
+
+<details>
+  <summary>Sample Request Body for Testing in Postman</summary>
+
+```JSON
+{
+    "name" : "Misty",
+    "personality" : "Bubbly",
+    "occupation" : "Gym Leader",
+    "age" : 25
+}
+
+```
+</details>
+</br>
+
+## Install the OpenAI Library and Connect Your Secret Key
+
+Before we write our request to OpenAI within the project, we need to install the OpenAI Library. 
+
+1. In your terminal, run `pip install --upgrade openai`
+2. Run `pip freeze > requirements.txt` to add the OpenAI library to your requirements.
+3. In your `.env` file, add the variable `OPENAI_API_KEY` and assign it the value of your secret key: `OPENAI_API_KEY=<Your Secret Key>`
+4. In the `character_routes.py` file, import OpenAI and create an OpenAI environment using the code below:
+    ```python
+
+    from openai import OpenAI
+
+    client = OpenAI()
+    ```
+| <div style="min-width:300px;">Part of Code</div> | Description |
+| ------------- | ----------- | 
+| `from openai import OpenAI` | Import the OpenAI library |
+| `client = OpenAI()` | This code creates an OpenAI environment with your secret key from the `.env` file. We will use `client` to make our API calls. |
+
+## Make the API call
+
+To make an API call, we want to go ahead and use the `client` that we created. We'll go ahead and nest our call inside a helper function called `generate_greetings` that will take a character in as a parameter and then use the character's attributes to construct a prompt for our request body. This may look something like this:
```

Comment:
```suggestion
To make an API call, we'll create a helper function called `generate_greetings` that will take a `Character` in as a parameter. Then we'll use the `Character`'s attributes to construct a prompt for our request body. This may look something like:
```